mono/mini/exceptions-arm64.c

```diff
@@ -620,3 +626,598 @@ mono_arch_do_ip_adjustment (MonoContext *ctx)
+	else
+		mono_arch_unwind_add_alloc_l (unwind_codes, unwind_code_size, size, reverse);
+}
+
+static int
+mono_arch_unwind_add_reg_offset (MonoUnwindOp* unwind_op_data, GSList* next, gint stack_offset, guint8 *unwind_codes, guint32 *unwind_code_size, gboolean reverse) {
+
+	if (unwind_op_data->reg == ARMREG_SP) {
+		g_assert("I don't believe there is a way to encode this in the unwind info");
+		return 0;
+	}
+
+	if (unwind_op_data->reg == ARMREG_FP || unwind_op_data->reg == ARMREG_LR) {
+		// FP and LR are handled seperately
+		return 0;
+	}
+
+	if (unwind_op_data->reg >= ARMREG_R19 && unwind_op_data->reg <= ARMREG_R28) {
+		// Only the presreved registers have unwind codes
+		// We shoulded hit here, but emit a nop if we do
+		g_assert(unwind_op_data->reg >= ARMREG_R19 && unwind_op_data->reg <= ARMREG_R28);
+		mono_arch_unwind_add_nop(unwind_codes, unwind_code_size);
+		return 1;
+	}
+
+	MonoUnwindOp* next_unwind_op_data = NULL;
+	guint offset = stack_offset + unwind_op_data->val;
+
+	if (next && next->data)
+		next_unwind_op_data = (MonoUnwindOp*)next->data;
+
+	if (next_unwind_op_data && unwind_op_data->reg % 2 && next_unwind_op_data->op == DW_CFA_offset && next_unwind_op_data->reg == unwind_op_data->reg + 1 && next_unwind_op_data->val == unwind_op_data->val + sizeof(host_mgreg_t)) {
+		if (offset)
+			mono_arch_unwind_add_save_regp_x(unwind_codes, unwind_code_size, unwind_op_data->reg, offset, reverse);
+		else
+			mono_arch_unwind_add_save_regp(unwind_codes, unwind_code_size, unwind_op_data->reg, offset, reverse);
+
+		return 2;
+	}
+
+	if (offset)
+		mono_arch_unwind_add_save_reg_x(unwind_codes, unwind_code_size, unwind_op_data->reg, offset, reverse);
+	else
+		mono_arch_unwind_add_save_reg(unwind_codes, unwind_code_size, unwind_op_data->reg, offset, reverse);
+
+	return 1;
+}
+
+static void
+mono_arch_unwind_add_emit_subx_sp_imm (gint offset, guint8* unwind_codes, guint32* unwind_code_size, gboolean reverse) {
+
+	/* Encodes emit_subx_sp_imm:
+		mov         xip0, offet & 0xFFFF
+		movk        xip0, #4, lsl offset >> 16 // if offset > 0xFFFF
+		mov         xip1, sp
+		add         xip1, xip1, xip0
+		mov	    sp, xip1
+	*/
+	mono_arch_unwind_add_nops(unwind_codes, unwind_code_size, offset > 0xFFFF ? 4 : 3);
+	mono_arch_unwind_add_alloc_x (unwind_codes, unwind_code_size, offset, reverse);
+}
+
+typedef struct _InitilizedWindowInfoResult {
+	gint32 saved_int_regs;
+	gboolean can_use_packed_format;
+	gboolean needs_seperate_epilog;
+} InitilizedWindowInfoResult;
+
+static InitilizedWindowInfoResult
+initialize_unwind_info_prolog (GSList* unwind_ops, gint stack_offset, guint param_area, guint8 *unwind_codes, guint32 *unwind_code_size) {
+
+	InitilizedWindowInfoResult res;
+	res.saved_int_regs = 0;
+	res.needs_seperate_epilog = FALSE;
+	res.can_use_packed_format = TRUE;
+
+	if (enable_ptrauth)
+		mono_arch_unwind_add_pac_sign_lr (unwind_codes, unwind_code_size);
+
+	// Stack probes
+	if (stack_offset + param_area > 0x1000) {
```

Comment:
This needs kept in sync with `emit_prolog_setup_sp_win64` in `mono/mini/mini-arm64.c`?

Reply:
Yes, basically all of https://github.com/Unity-Technologies/mono/pull/1960/files#diff-90c089fd795716a5ad3b86b5d0fcb22523e846ffd2d286f8189d7e72e081dedcR862-R895 has to be kept in sync.  

I could add more special unwind codes - like `W_CFA_mono_fp_alloc_info_win64_arm64` to track this.  There could be some ordering issues, but it's probably worthwhile

Reply:
And this we already broke by https://github.com/Unity-Technologies/mono/pull/1965

I've fixed this in https://github.com/Unity-Technologies/mono/pull/1960/commits/adb3c97d4d97ce929bdf1f88eeb6eba5d0f17b18.  It's somewhat awkward in that the frame setup gets processed in the unwind code loop.  But this lets us just emit nop's for the stack probes.

@joncham This is a big enough change that I'm requesting a re-review.